### PR TITLE
 Pass the userInfo dictionary as an argument to the app state handlers #495

### PR DESCRIPTION
**Pull request by Wenzhi**
This is critical for detecting push opens when the application does a cold-start (i.e. from suspended/closed) from a push click and the push-clicked signal (i.e. push payload in the launchOptions) is in the userInfo.

### CODE REVIEW

**Comment by f2prateek** [comment]
Pretty sure this doesn't actually work. Have you tested this?

**Comment by Wenzhi** [comment]
@f2prateek I tested on iOS 9.2 on a 6s Plus, and the launch dictionary is passed down to `SEGAppboyIntegration`'s method `- (void)applicationDidFinishLaunching:(NSDictionary *)launchOptions`.

I can send you a sample project if you want.

**Comment by f2prateek** [comment]
Interesting, I figured it wouldn't work because the method definition is different: https://github.com/segmentio/analytics-ios/blob/master/Pod/Classes/Integrations/SEGIntegration.h#L62. It doesn't accept an NSDictionary, but it accepts NSNotification.

**Comment by f2prateek** [comment]
@Wenzhi could you explain  how the `- (void)applicationDidFinishLaunching:(NSDictionary *)launchOptions` was being invoked even though the method signature is different https://github.com/segmentio/analytics-ios/blob/master/Pod/Classes/Integrations/SEGIntegration.h#L62

**Comment by Wenzhi** [comment]
Hey @f2prateek,

I originally made the method accept an NSDictionary by following the scheme of your callbacks for remote notifications (it worked because objective C will call that method if it exists regardless of what the protocol says).  I’ve updated the pull to correct the method signature.  I’ve also updated to keep didFinishLaunching running until the SDKs finish launching. By doing so, we can collect the push statistics and handle the HTTP-links/deep-links in the push payload.  A lot of push detection and handling depends on application state, and without this, the signal gets lost.

**Comment by f2prateek** [comment]
I have to follow up questions:

1. We're already calling this method synchronously on the integration so you should get the correct application state — why do we need to keep the `didFinishLaunching` method running in addition to this?

2. What application state are you recording? Can you share a concrete example so I can understand better?

**Comment by Wenzhi** [comment]
Hey @f2prateek,

The issue we’re seeing is that the applicationDidFinishLaunching callback seems to not be getting called synchronously - it’s happening before initialized is set to true, so it’s doing it async.  We need to look at application state because the push signal in applicationDidFinishLaunching (i.e. a push payload) looks the same when the app is launched in the background and when the app is being launched into the foreground.  We need the app state to determine if it’s a push open or a background delivery.(edited)

Furthermore, we route behavior through shared push handling code, which relies on similar application state signal (in relation to didReceiveRemoteNotification calls).  If didFinishLaunching is called from an Active state, it looks like a push received in the foreground (which we log differently than a regular push because the system doesn’t show it)

**Comment by f2prateek** [comment]
I think we might have to chalk this down to a limitation as a segment integration. We'll try to deliver it synchronously when we can, but obviously it's not always possible, and I'd rather not block the lifecycle methods for this.

**Comment by Wenzhi** [comment]
How about only keep the method running when Appboy is intergrated?
```
if ([note.name isEqualToString:UIApplicationDidFinishLaunchingNotification] && 
    NSClassFromString(@"SEGAppboyIntegrationFactory") != nil) {
+      while (!self.initialized) {
+        [[NSRunLoop currentRunLoop] runMode:NSDefaultRunLoopMode beforeDate:[NSDate dateWithTimeIntervalSinceNow: 1.0]];
+      }
+    }
```

**Comment by f2prateek** [comment]
That wouldn't be better since then we'd have integration specific code in the core library.

The main problem here is the 1s hardcoded duration. What guarantees that Appboy (and any other integration needs this) gets the event within a second and processes it? it could easily be outside this window and we'd back to square one.

**Comment by Wenzhi** [comment]
Hey Prateek, that code is in a while loop, so it'll keep waiting till `self.initialized` is YES (checking in 1 second intervals).  I think that should be a fairly clean and efficient way to wait for everything to be set up before calling the integrations - please let me know what you think.

**Comment by f2prateek** [comment]
But in that case, what happens if the integrations never get initialized? This scenario is not unlikely if the user's device is offline on the very first time they launch the app — we never get any settings and don't have anything cached so we're blocking continuously and indefinitely.

File: Pod/Classes/SEGAnalytics.m [hunk 1]
```diff
@@ -158,8 +158,19 @@ - (void)handleAppStateNotification:(NSNotification *)note
                             };
     });
     SEL selector = NSSelectorFromString(selectorMapping[note.name]);
-    if (selector)
-        [self callIntegrationsWithSelector:selector arguments:nil options:nil sync:true];
+  if (selector) {
```
Comment (f2prateek) [inline Pod/Classes/SEGAnalytics.m]: Indentation here is slightly off (I can fix this later as well).

File: Pod/Classes/SEGAnalytics.m [hunk 2]
```diff
@@ -158,8 +158,19 @@ - (void)handleAppStateNotification:(NSNotification *)note
...
+    if ([note.name isEqualToString:UIApplicationDidFinishLaunchingNotification]) {
+      while (!self.initialized) {
+        [[NSRunLoop currentRunLoop] runMode:NSDefaultRunLoopMode beforeDate:[NSDate dateWithTimeIntervalSinceNow: 1.0]];
+      }
+    }
+    NSArray *arguments = nil;
+    if (note != nil) {
+      arguments = @[note];
```
Comment (f2prateek) [inline Pod/Classes/SEGAnalytics.m]: Should we update the definitions here to indicate that a dictionary might be passed in as an argument? https://github.com/segmentio/analytics-ios/blob/master/Pod/Classes/Integrations/SEGIntegration.h#L62-L67

**Comment by f2prateek** [comment]
Last 2 questions, lgtm to otherwise!

File: Pod/Classes/SEGAnalytics.m (hunk 1 above)
Comment (Wenzhi) [inline Pod/Classes/SEGAnalytics.m]: I fixed it. Thank you for catching it.

File: Pod/Classes/SEGAnalytics.m (hunk 2 above)
Comment (Wenzhi) [inline Pod/Classes/SEGAnalytics.m]: I think those definitions are fine. I updated my code so a notification is passed in instead of a dictionary.

Comment (f2prateek) [inline Pod/Classes/SEGAnalytics.m]: Hey @Wenzhi, I think the code is still the same. Mind updating that and we can merge this and make a release?

Comment (f2prateek) [inline Pod/Classes/SEGAnalytics.m]: Ah nvm I see what you meant.

**Comment by f2prateek** [comment]
Sorry @Wenzhi I wanted to make a release today so I undid the PR for now, could you re-open it?

**Pull request merged**